src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/PackageFeeds/MultiSourcePackageMetadataProvider.cs

```diff
@@ -193,6 +197,27 @@ private async Task<IEnumerable<VersionInfo>> FetchAndMergeVersionsAsync(PackageI
             return await MergeVersionsAsync(identity, completed);
         }
 
+        private async Task<PackageDeprecationMetadata> FetchDeprecationMetadataAsync(PackageIdentity identity, CancellationToken cancellationToken)
+        {
+            // Take the deprecation metadata from the first source it is found in
+            foreach (var source in _sourceRepositories)
+            {
+                var metadata = await source.GetPackageMetadataAsync(identity, true, cancellationToken);
+                if (metadata == null)
+                {
+                    continue;
```

Comment:
As I understand it, this method will serially query each package source until it finds deprecation metadata or until it has queried all package sources, and this method will be called for each package.

This sure seems like a potential perf issue and an opportunity for caching.

Is this metadata cached client-side?

Ideally, this code would have some package source preference (e.g.:  implicitly known from wherever the package was installed or by some global preference).  Some users have many packages sources (+10), so I can't imagine iterating through them all for each package would be a generally good idea.

Reply:
I don't believe caching should be at this level--IMO the `SourceRepository` or `PackageMetadataResource` should be the ones caching, and the UI infrastructure shouldn't have to maintain its own cache.

As far as whether or not it is caching, it appears not to be (and I believe @nkolev92 told me explicitly that this was intended) as it makes HTTP calls to registration every time the PM UI is loaded or refreshed.

Regarding the fact that this method queries each package source, that was actually unintended at first, and was just a small bug in that original implementation. First, I was able to update the method so that it asynchronously called all sources, and would return the deprecation metadata from the first feed that had the package. Then, however, I looked into it some more, and it seems that the call to fetch the list of versions for the package currently does enumerate all package source and stitch together the full list of versions from it. I was able to rewrite the code so that it reuses the same calls to registration as the fetching of the version list, so the fetching of deprecation metadata shouldn't have any additional impact on the PM UI's perf.

Does this resolve your concern?